### DISCUSSION TITLE
Lockfile update for Azure Artifacts on self-hosted DevOps Server 2022 not supported

### DISCUSSION THREAD

**Discussion by sbaeurle** [renovatebot/renovate#25365] [Request Help] [answered]
### How are you running Renovate?

Self-hosted Renovate

### If you're self-hosting Renovate, tell us what version of Renovate you run.

37.31.4

### If you're self-hosting Renovate, select which platform you are using.

Azure DevOps Server

### Was this something which used to work for you, and then stopped?

I am trying to get this working for the first time

### Describe the problem

When updating packages.lock.json files using renovate in combination with Azure DevOps server configured to utilze NTLM as possible authentication mechanism, the issue dotnet restore command fails with the following error:

```
GSSAPI operation failed with error - An invalid name was supplied (Configuration file does not specify default realm)
```
This is due to a similar issue to: https://stackoverflow.com/questions/59585950/getting-error-during-dotnet-restore-in-docker-container

In order to make this work, one has to add `<add key="ValidAuthenticationTypes" value="basic" />` into the credential section of the `nuget.config`.
The corresponding code snippet would be: https://github.com/renovatebot/renovate/blob/8be845a0839bfb148be226952e13d40609bcd96d/lib/modules/manager/nuget/config-formatter.ts#L86C10-L104

How could one make this configurable? I'm more than happy to create a corresponding PR myself, just need some input on how to structure the fix to best fit renovate guidelines :)

### Relevant debug logs

<details><summary>Logs</summary>

{
  "name": "renovate",
  "hostname": "50a869275303",
  "pid": 192,
  "level": 20,
  "logContext": "WfTzcg8UXU6nl0cC1evLU",
  "repository": "*.git",
  "branch": "chores/dotnet-monorepo",
  "baseBranch": "bugs/lockfile-update-fails",
  "err": {
    "cmd": "/bin/sh -c dotnet restore Backend/Tests/UnitTests/UnitTests.csproj --force-evaluate --configfile /tmp/renovate/cache/__renovate-private-cache/nuget/nuget.config",
    "stderr": "",
    "stdout": "\nWelcome to .NET 6.0!\n---------------------\nSDK Version: 6.0.123\n\n----------------\nInstalled an ASP.NET Core HTTPS development certificate.\nTo trust the certificate run 'dotnet dev-certs https --trust' (Windows and macOS only).\nLearn about HTTPS: https://aka.ms/dotnet-https\n----------------\nWrite your first app: https://aka.ms/dotnet-hello-world\nFind out what's new: https://aka.ms/dotnet-whats-new\nExplore documentation: https://aka.ms/dotnet-docs\nReport issues and find source on GitHub: https://github.com/dotnet/core\nUse 'dotnet --help' to see available commands or visit: https://aka.ms/dotnet-cli\n--------------------------------------------------------------------------------------\n  Determining projects to restore...\n/usr/lib/dotnet/sdk/6.0.123/NuGet.targets(130,5): error : Unable to load the service index for source https://<devop-url>/_packaging/<feed-name>/nuget/v3/index.json. [/tmp/renovate/repos/azure/*/UnitTests.csproj]\n/usr/lib/dotnet/sdk/6.0.123/NuGet.targets(130,5): error :   GSSAPI operation failed with error - An invalid name was supplied (Configuration file does not specify default realm). [/tmp/renovate/repos/azure/*/UnitTests.csproj]\n",
    "options": {
      "cwd": "/tmp/renovate/repos/azure/*.git",
      "encoding": "utf-8",
      "env": {
        "NUGET_PACKAGES": "/tmp/renovate/cache/__renovate-private-cache/nuget/packages",
        "MSBUILDDISABLENODEREUSE": "1",
        "HOME": "/root",
        "PATH": "/opt/microsoft/powershell/7:/opt/node-v18.18.2-linux-x64/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
        "CONTAINERBASE_CACHE_DIR": "/tmp/renovate/cache/containerbase"
      },
      "maxBuffer": 10485760,
      "timeout": 900000
    },
    "exitCode": 1,
    "name": "ExecError",
    "message": "Command failed: dotnet restore */UnitTests.csproj --force-evaluate --configfile /tmp/renovate/cache/__renovate-private-cache/nuget/nuget.config\n",
    "stack": "ExecError: Command failed: dotnet restore Backend/Tests/UnitTests/UnitTests.csproj --force-evaluate --configfile /tmp/renovate/cache/__renovate-private-cache/nuget/nuget.config\n\n    at ChildProcess.<anonymous> (/opt/node-v18.18.2-linux-x64/lib/node_modules/renovate/lib/util/exec/common.ts:99:11)\n    at ChildProcess.emit (node:events:529:35)\n    at ChildProcess.emit (node:domain:489:12)\n    at Process.ChildProcess._handle.onexit (node:internal/child_process:292:12)"
  },
  "durationMs": 2976,
  "msg": "rawExec err",
  "time": "2023-10-23T09:12:15.599Z",
  "v": 0
}

</details>

### Have you created a minimal reproduction repository?

I have linked to a minimal reproduction in the description above

**Comment by sbaeurle** [comment]
Suggested changes can be found here: https://github.com/sbaeurle/renovate/commit/8e8bfd12ca2782ce20ce52e214277c7e794be02f

**Reply by rarkins** [reply]
If it is based on authType then maybe it doesn't need to be configurable - it should be default behavior. But I think it should perhaps be enabled if authType=Basic instead of if authType is defined at all as you did in your commit?

**Reply by sbaeurle** [reply]
From my understanding of https://docs.renovatebot.com/configuration-options/#authtype this could cause some problems however.
So if I set `authType: Basic` in my renovate.json, the bot would ignore username and password set in the config right?

**Reply by rarkins** [reply]
Please describe your question in more detail as I don't understand it

**Reply by sbaeurle** [reply]
If I set `authType: Basic` in the Renovate config, it seems that (see [here](https://docs.renovatebot.com/configuration-options/#authtype:~:text=Do%20not%20set%20authType%3DBearer%3A%20it%27s%20the%20default%20setting%20for%20Renovate%20anyway.%20Do%20not%20set%20a%20username%20or%20password%20when%20you%27re%20using%20authType%2C%20as%20authType%20doesn%27t%20use%20usernames%20or%20passwords.)) the username and password fields are ignored during the dependency lookup.
This would mean I have to set `password` and `username` for the dotnet restore and `token: <Basic-Auth-Base64-String>` in the configuration.

Seems rather counterintuitive to me so we should potentially use a different approach.

**Reply by sbaeurle** [reply]
I did take a look at the Nuget Configuration page of the renovate docs and since only Basic Authentication is supported anyway, it's probably best to just add the `ValidAuthenticationTypes` as default behavior.
Will create a PR for it

**Comment by sbaeurle** [comment, marked as answer]
Solved in #25502
